external-crates/move/documentation/book/src/control-flow/pattern-matching.md

```diff
@@ -0,0 +1,567 @@
+        | <variable>
+        | _
+        | C { inner-pattern, inner-pattern, ... } // where C is a struct or enum variant
+        | C ( inner-pattern, inner-pattern, ... ) // where C is a struct or enum variant
+        | C                                       // where C is an enum variant
+        | <variable> @ top-level-pattern
+        | pattern | pattern
+inner-pattern = pattern
+              | ..
+              | mut pattern
+```
+
+Patterns that contain variables bind them to the value being matched. These variables can then be
+used either in any match guard expressions, or on the right-hand side of the match arm. For example:
+
+```move
+public struct Wrapper(u64)
+
+fun add_under_wrapper_unless_equal(wrapper: Wrapper, x: u64): u64 {
+    match (wrapper) {
+        Wrapper(y) if (y == x) => Wrapper(y),
+        Wrapper(y) => y + x,
+    }
+}
+add_under_wrapper_unless_equal(Wrapper(1), 2); // returns Wrapper(3)
+add_under_wrapper_unless_equal(Wrapper(2), 3); // returns Wrapper(5)
+add_under_wrapper_unless_equal(Wrapper(3), 3); // returns Wrapper(3)
+```
+
+Patterns can be nested, and patterns can be or'd with other patterns. The `..` pattern is a special
+pattern that matches any number of fields in a struct or enum variant, but it can only occur within
+a constructor pattern, similarly the `mut` pattern can only be used within constructor patterns --
+this is used to specify that we want to use the variable mutably on the right-hand-side of the match
+arm.
+
+Patterns are not expressions, but they are nevertheless typed just like expressions. This means that
+the type of a pattern must match the type of the value it matches. For example, the pattern `1` has
+type `u64`, the pattern `MyEnum::Variant(1, true)` has type `MyEnum`, and the pattern
+`MyStruct { x, y }` has type `MyStruct`. If you try to match on an expression which differs from the
+type of the pattern in the match this will result in a type error. For example:
+
+```move
+match (1) {
+    // The `true` literal pattern is of type `bool` so this is a type error.
+    true => 1,
+    // TYPE ERROR: expected type u64, found bool
+    _ => 2,
+}
+```
+
+Similarly the following would also result in a type error since `MyEnum` and `MyStruct` are
+different types:
+
+```
+match (MyStruct { x: 0, y: 0 }) {
+    MyEnum::Variant(..) => 1,
+    // TYPE ERROR: expected type MyEnum, found MyStruct
+}
+```
+
+Additionally, there are some restrictions on when the `..` pattern, and `mut` pattern modifier can
+be used in a pattern.
+
+A `mut` modifier can only occur within a constructor pattern, and cannot be a top-level pattern. The
+value being matched on must be either a mutable reference or by value in order for a `mut` pattern
+to be used otherwise the compiler will raise an error.
+
+```move
+public struct MyStruct(u64)
+
+fun top_level_mut(x: MyStruct) {
+    match (x) {
+        mut MyStruct(y) => 1,
+        // ERROR: cannot use mut pattern as a top-level pattern
```

Comment:
Decide if the `// ERROR` comment goes above or below and make it consistent across the file